radicle-node/src/wire/protocol.rs

```diff
@@ -0,0 +1,318 @@
+            hangups: HashMap::default(),
+        }
+    }
+
+    fn disconnect(&mut self, fd: RawFd, reason: DisconnectReason<service::DisconnectReason>) {
+        let node_id = self
+            .sessions
+            .remove(&fd)
+            .expect("disconnecting unknown peer");
+        log::debug!("Attempting to disconnect {} due to {}", node_id, reason);
+        self.hangups.insert(fd, Some(reason));
+        self.inner_queue
+            .push_back(reactor::Action::UnregisterTransport(fd));
+    }
+}
+
+impl<R, S, W, G> reactor::Handler for Transport<R, S, W, G>
+where
+    R: routing::Store + Send,
+    S: address::Store + Send,
+    W: WriteStorage + Send + 'static,
+    G: Signer + Negotiator + Send,
+    <G as Ecdh>::Pk: Send,
+{
+    type Listener = NetAccept<Session<G>>;
+    type Transport = NetTransport<Session<G>, Message>;
+    type Command = service::Command;
+
+    fn handle_wakeup(&mut self) {
+        self.service.wake()
+    }
+
+    fn handle_listener_event(
+        &mut self,
+        socket_addr: net::SocketAddr,
+        event: ListenerEvent<Session<G>>,
+        duration: Duration,
+    ) {
+        self.service.tick(LocalTime::from_secs(duration.as_secs()));
+
+        match event {
+            ListenerEvent::Accepted(session) => {
+                log::debug!(
+                    "Incoming connection from remote peer {}",
+                    session.transition_addr()
+                );
+                self.handshakes.insert(session.as_raw_fd(), Link::Inbound);
+                let transport = NetTransport::<Session<G>, Message>::upgrade(session)
+                    .expect("socket failed configuration");
+                self.service.inbound(socket_addr);
+                self.inner_queue
+                    .push_back(reactor::Action::RegisterTransport(transport))
+            }
+            ListenerEvent::Error(err) => {
+                panic!("I/O error on the listener socket, {:?}", err)
+            }
+        }
+        // TODO: Ensure we do not need to generate any events or do calls to the inner
+    }
+
+    fn handle_transport_event(
+        &mut self,
+        fd: RawFd,
+        event: SessionEvent<Session<G>, Message>,
+        duration: Duration,
+    ) {
+        self.service.tick(LocalTime::from_secs(duration.as_secs()));
+
+        match event {
+            SessionEvent::SessionEstablished(node_id) => {
+                if let Some((fd, _)) = self.sessions.iter().find(|(_, id)| **id == node_id) {
+                    log::warn!(
+                        "New session from the same peer {}, closing previous session",
+                        node_id
+                    );
+                    self.disconnect(
+                        *fd,
+                        DisconnectReason::DialError(
+                            io::Error::from(io::ErrorKind::AlreadyExists).into(),
+                        ),
+                    );
+                    return;
```

Comment:
Is it possible that more than one session exists here for the same NodeId, that needs to be disconnected?

Reply:
The rule is "one session per node id" always. Since it doesn't make sense otherwise and the proper business logic can't be constructed without that assumption (each session has a different noise encoder state, and for outcoming messages addressing some NodeId it can't be decided which session to use).

Thus, if a new connection happens with the peer with the same node id, either the connection must not happen (possible for outgoing connections) or the previous session has to be closed (for incoming connection) at the moment the handshake is done.

Again, github UI doesn't show me the code around you comment, so if it is not the case of what's happened, I need to do a fix.